pinot-server/src/main/java/org/apache/pinot/server/api/resources/SegmentMetadataFetcher.java

```diff
@@ -77,8 +80,13 @@ public static String getSegmentMetadata(SegmentDataManager segmentDataManager, L
       if (immutableSegment instanceof ImmutableSegmentImpl) {
         ImmutableSegmentImpl immutableSegmentImpl = (ImmutableSegmentImpl) immutableSegment;
         Map<String, ColumnIndexContainer> columnIndexContainerMap = immutableSegmentImpl.getIndexContainerMap();
-        columnIndexMap = getImmutableSegmentColumnIndexes(columnIndexContainerMap);
+        columnIndexMap = getSegmentColumnIndexes(columnIndexContainerMap);
```

Comment:
Instead of all this type-casting, why not extend the SegmentDataManager interface to include a method to get column indices?

Reply:
Agreed, lot of type casting going on here. How about have it in the IndexSegment interface instead? Since managing the IndexContainer is really the job of the segment than the SegmentDataManager.

Reply:
I am fine with that too. @mayankshriv is changing the way all this is organized, so we might as well get to a final state that he has in mind.